plugins/metrics/pkg/cortex/api_types.go

```diff
@@ -0,0 +1,124 @@
+package cortex
+
+import (
+	"regexp"
+
+	"golang.org/x/exp/slices"
+)
+
+func xOR(expr1, expr2 bool) bool {
+	return expr1 != expr2
+}
+
+type listRuleFilter struct {
```

Comment:
Do we really need a new type here? Seems like there's some coed duplication in `NewListRuleFilter` and the validation function `func (l *ListRulesRequest) Validate() error` in `plugins/metrics/pkg/apis/cortexadmin/extensions.go`.

Can we refactor this to a couple methods instead?

```go
func MatchesCluster(request *cortexadmin.ListRulesRequest, clusterId string) bool {
	if *request.RequestAll {
		return true
	}

	// we store each cluster rules in a file with the cluster id
	// and when we search invalid we search specifically those that do not match existing clusters
	return xOR(*request.ListInvalid, slices.Contains(request.ClusterId, clusterId))
}
```

LGTM otherwise

Reply:
Yep